### PR TITLE
Improved zerobufCxx.py for ipp files generation

### PR DESCRIPTION
**Pull request by gonkc**
Hello, I'm Gonzalo Bayo, from the group working back in URJC/UPM with Pablo Toharia. Right now I'm trying to use **ZeroBuf** in a custom C++ project in Windows and the generated ipp files don't compile if we use this library in more than one project. Visual Studio compiler reject the compilation with an LNK2005( function/method already defined ).
Modifying the _zerobufCxx.py_ file by adding the "inline" keyword before function definitions on ipp files could be an option to fix this and make errors disappear.
Regards

### CODE REVIEW

File: bin/zerobufCxx.py [hunk 1]
```diff
@@ -198,14 +198,20 @@ def write_declaration(self, file):
...
+        
         if self.ret_val: # '{}'-less body
-            file.write('{0} {1}{2}\n{{'.format(self.ret_val,
-                (classname + '::') if classname else '', impl_function))
+            if self.ret_val.startswith("template<>") or has_final:
+                file.write('{0} {1}{2}\n{{'.format(self.ret_val,
+                    (classname + '::') if classname else '', impl_function))
+            else:
+                file.write('inline {0} {1}{2}\n{{'.format(self.ret_val,
+                    (classname + '::') if classname else '', impl_function))
```
Comment (gonkc) [inline bin/zerobufCxx.py]: Fixed with a new parameter on format string.

**Comment by rdumusc** [comment]
looks OK to me, but we have to remove shallow = true from .gitmodules to make CI pass

**Comment by gonkc** [comment]
We've getting some errors making the compilation with cpp files with 'inline' keyword. To solve that, 'inline' keyword only is used on inline implementations (ipp files).
Regards

File: bin/zerobufCxx.py [hunk 2]
```diff
@@ -15,6 +15,8 @@
 import sys
 from collections import namedtuple, OrderedDict
 
+inline_implementation = False
```
Comment (rdumusc) [inline bin/zerobufCxx.py]: Unless there is a good reason for this global "option" I would remove it and always add inline

**Review by rdumusc — CHANGES_REQUESTED** [review]
Thanks for the fix. Please rebase your PR on https://github.com/HBPVIS/ZeroBuf/pull/85 to fix the CI build, which is needed for merging.

File: bin/zerobufCxx.py (hunk 2 above)
Comment (gonkc) [inline bin/zerobufCxx.py]: I think that it won't work. In Linux these functions are not inline and I would not force the use of inline functions.

Comment (tribal-tec) [inline bin/zerobufCxx.py]: Please check https://github.com/HBPVIS/ZeroBuf/commit/7ab0058e24cc262eb95a27a344b4a6bee78429cf; there is already enough code to decide if to use inline on Windows, and not-inline everywhere else. I wonder why I forgot the inline back then, it seemed to have worked, but that's obviously not working.

Comment (gonkc) [inline bin/zerobufCxx.py]: Removed global default value of inline_implementation and global keyword. Only needed read access to variable.

File: bin/zerobufCxx.py [hunk 3]
```diff
@@ -198,14 +198,22 @@ def write_declaration(self, file):
...
         impl_function = re.sub(r" = [0-9\.f]+ ", " ", impl_function) # remove default params
 
         next_line(file)
+        
         if self.ret_val: # '{}'-less body
-            file.write('{0} {1}{2}\n{{'.format(self.ret_val,
-                (classname + '::') if classname else '', impl_function))
+            add_before_inline = ""
+            add_after_inline = ""
+            if inline_implementation is True:
```
Comment (tribal-tec) [inline bin/zerobufCxx.py]: no need to compare against 'True'

**Review by dnachbaur — APPROVED** [review]
Good otherwise. Thx for the fix!

File: bin/zerobufCxx.py (hunk 3 above)
Comment (gonkc) [inline bin/zerobufCxx.py]: That's right, I forgot to remove it. Now it's fixed. Thanks.

File: bin/zerobufCxx.py (hunk 2 above)
Comment (gonkc) [inline bin/zerobufCxx.py]: @rdumusc , review this change, please. Thanks.

**Review by rdumusc — APPROVED** [review]

**Comment by rdumusc** [comment]
retest this please

**Pull request merged**